psh/test-ls.py

```diff
@@ -0,0 +1,153 @@
+        pattern += r'(\x1b\[\S*)?' + f'{file}' + r'((\s*)|(\x1b\[\S*))+' + new_line_reg
+    if not separate_lines:
+        pattern += psh.EOL
+    return pattern
+
+
+def assert_ls_noarg(p, files):
+    expected = create_pattern(sorted(files))
+    msg = 'Wrong content of listed directory, when calling ls without arguments'
+    psh.assert_cmd(p, 'ls test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_1(p, files):
+    expected = create_pattern(sorted(files), separate_lines=True)
+    msg = 'Wrong content of listed directory, when calling `ls -1`'
+    psh.assert_cmd(p, 'ls -1 test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_a(p, files):
+    expected = create_pattern(sorted(files))
+    msg = 'Wrong content of listed directory, when calling `ls -a`'
+    psh.assert_cmd(p, 'ls -a test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_d(p):
+    expected = r'(\x1b\[\S*)?' + 'test_ls_dir' + r'\s*' + r'(\x1b\[\S*)?' + r'(\r+\n)?'
+    msg = 'Wrong output, when calling `ls -d tested_directory`'
+    psh.assert_cmd(p, 'ls -d test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_d_noarg(p):
+    expected = r'(\x1b\[\S*)?' + '.' + r'\s*' + r'(\x1b\[\S*)?' + r'(\r+\n)?'
+    msg = 'Wrong output, when calling `ls -d` without specified directory'
+    psh.assert_cmd(p, 'ls -d', expected, msg, is_regex=True)
+
+
+def assert_ls_f(p, files):
+    expected = create_pattern(files)
+    msg = 'Wrong output, when calling `ls -f test_ls_dir`'
+    psh.assert_cmd(p, 'ls -f test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_h(p):
+    expected = r'usage:(\s)ls \[options\] \[files\](\r+\n)' + \
+                r'(\s*-[\w\s]:[ \S]+(\r+\n))+'
+    msg = 'Wrong ls help message format'
+    psh.assert_cmd(p, 'ls -h', expected, msg, is_regex=True)
+
+
+def assert_ls_l(p, files):
+    cmd = 'ls -l test_ls_dir'
+    p.sendline(cmd)
+    p.expect(cmd)
+    for file in sorted(files):
+        if file[0] == '.':
+            continue
+        p.expect(r'(-|b|c|d|l|n|p|s|D|E|O|S)([-r][-w][-x]){3}( )'
+                 + r'(\d)( )'
+                 + r'(root |--- ){2}( *)'
+                 + r'(\d+)( )'
+                 + r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)( )'
+                 + r'[1-31]( *)'
+                 + r'(\d){2}(:)(\d){2}( )'
+                 + r'(\x1b\[34m)?'
+                 + f'{file}')
+
+    psh.assert_prompt(p, "Prompt hasn't been seen after calling `ls -l test_ls_dir`")
+
+
+def assert_ls_r(p, files):
+    expected = create_pattern(sorted(files, reverse=True))
+    msg = 'Wrong output, when calling `ls -r test_ls_dir`'
+    psh.assert_cmd(p, 'ls -r test_ls_dir', expected, msg, is_regex=True)
+
+
+def assert_ls_S(p, files):
+    size_sorted_files = list(sorted(files))
+    dir_index = size_sorted_files.index('dir')
+    size_sorted_files[0], size_sorted_files[dir_index] = size_sorted_files[dir_index], size_sorted_files[0]
```

Comment:
This is a noop for files used in harness as dir would be first after sorting anyway. This could create invalid order in other cases:

`('a', 'b', 'dir')` would become `('dir', 'b', 'a')`


Key parameter can be used to modify sorting rules:
```py
size_sorted_files = list(sorted(files, key=lambda x: (0, x) if is_dir(x) else (1, x)))
```
sorted function in this example will compare first member of tuple first so directories are given "higher priority" here.

Reply:
You are right, corrected.